src/transactions/TransactionFrame.h

```diff
@@ -240,68 +197,77 @@ class TransactionFrame : public TransactionFrameBase
                            std::optional<int64_t> baseFee,
                            bool applying) const override;
 
-    void addSignature(SecretKey const& secretKey);
-    void addSignature(DecoratedSignature const& signature);
-
     bool checkSignature(SignatureChecker& signatureChecker,
-                        LedgerTxnEntry const& account, int32_t neededWeight);
+                        LedgerTxnEntry const& account,
+                        int32_t neededWeight) const;
 
     bool checkSignatureNoAccount(SignatureChecker& signatureChecker,
-                                 AccountID const& accountID);
-    bool checkExtraSigners(SignatureChecker& signatureChecker);
-
-    bool checkValidWithOptionallyChargedFee(Application& app,
-                                            AbstractLedgerTxn& ltxOuter,
-                                            SequenceNumber current,
-                                            bool chargeFee,
-                                            uint64_t lowerBoundCloseTimeOffset,
-                                            uint64_t upperBoundCloseTimeOffset);
-    bool checkValid(Application& app, AbstractLedgerTxn& ltxOuter,
-                    SequenceNumber current, uint64_t lowerBoundCloseTimeOffset,
-                    uint64_t upperBoundCloseTimeOffset) override;
-    bool checkSorobanResourceAndSetError(Application& app,
-                                         uint32_t ledgerVersion) override;
+                                 AccountID const& accountID) const;
+    bool checkExtraSigners(SignatureChecker& signatureChecker) const;
+
+    std::pair<bool, MutableTxResultPtr> checkValidWithOptionallyChargedFee(
+        Application& app, AbstractLedgerTxn& ltxOuter, SequenceNumber current,
+        bool chargeFee, uint64_t lowerBoundCloseTimeOffset,
+        uint64_t upperBoundCloseTimeOffset) const;
+    std::pair<bool, MutableTxResultPtr>
+    checkValid(Application& app, AbstractLedgerTxn& ltxOuter,
+               SequenceNumber current, uint64_t lowerBoundCloseTimeOffset,
+               uint64_t upperBoundCloseTimeOffset) const override;
+    bool
+    checkSorobanResourceAndSetError(Application& app, uint32_t ledgerVersion,
+                                    MutableTxResultPtr txResult) const override;
+
+    MutableTxResultPtr createTxResult() const override;
```

Comment:
Shall we name this `createSuccessResult` (and `createSuccessResultWithFeeCharged`) in order to better convey the semantics of the returned result?

FWIW it could be safer to create an 'uninitialized' result, so that it's not possible to accidentally have a successful result due early return, but that could be done as a followup.